Expose Cadence Code Coverage through Flow Emulator

m-Peter opened the pull request:
**Closes** https://github.com/onflow/flow-emulator/issues/333
**Depends On** https://github.com/onflow/flow-go/pull/4126

## Description

 Allow the Flow Emulator to utilize the built-in code coverage functionality from Cadence, and expose it to developers through:
- API (programmatic)
- HTTP endpoint (visual)
______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the GitHub PR explorer
- [x] Added appropriate labels

bluesign commented on blockchain.go:
```diff
@@ -612,11 +619,6 @@ func (b *Blockchain) newFVMContextFromHeader(header *flowgo.Header) fvm.Context
 	return fvm.NewContextFromParent(
 		b.vmCtx,
 		fvm.WithBlockHeader(header),
-		fvm.WithReusableCadenceRuntimePool(
-			reusableRuntime.NewReusableCadenceRuntimePool(
-				1,
-				runtime.Config{Debugger: b.debugger, AccountLinkingEnabled: true}),
-		),
```
is there a reason to remove this? this should be needed for debugger to function ( not sure though, flow-go is changing too fast )

m-Peter replied:
I had to remove this line here, because it would create a new `CoverateReport` object, which would be ephemeral during script & transaction execution.

The `ReusableCadenceRuntimePool` is properly initialized here: https://github.com/onflow/flow-emulator/blob/master/blockchain.go#L427-L446, so with:

```go
return fvm.NewContextFromParent(
    b.vmCtx,
    fvm.WithBlockHeader(header),
)
```

the `b.vmCtx` contains all the necessary config for the `Debugger`, `AccountLinkingEnabled` and `CoverageReportingEnabled` from the initial setup.

bluesign replied:
ah you have right, I missed that

bluesign commented on server/emulator.go:
```diff
@@ -65,6 +65,8 @@ func NewEmulatorAPIServer(server *EmulatorServer, backend *backend.Backend, stor
 
 	router.HandleFunc("/emulator/config", r.Config)
 
+	router.HandleFunc("/emulator/codeCoverage", r.CodeCoverage).Methods("GET")
+
 	return r
```
nit: would it be possible to also add a method to reset coverage report ? It can be very useful.

m-Peter replied:
That is quite a nice idea :100: :100: :100: 
I will add it for sure!!!

m-Peter replied:
@bluesign With the current API, we can achieve this with something like:

```go
func (b *Blockchain) ResetCoverageReport() {
	b.vmCtx.ReusableCadenceRuntimePool.CoverageReport = runtime.NewCoverageReport()
	b.coverageReport = b.vmCtx.ReusableCadenceRuntimePool.CoverageReport
}
```

I am not a big fan of reaching this with `b.vmCtx.ReusableCadenceRuntimePool.CoverageReport`, however, the `CoverageReport` lacks a `Reset` method, for the time being.

What are your thoughts on this? Do we first add the required API in `CoverageReport`, or we add a workaround like above, and come back to fix it later on?

m-Peter replied:
I have added here (https://github.com/onflow/flow-emulator/pull/332/commits/4e7b422e36c101f156ae74efa40ca262bd0958f6) the workaround, for reference.

bluesign commented on blockchain.go:
```diff
@@ -207,6 +210,7 @@ var defaultConfig = func() config {
 		Logger:                       zerolog.Nop(),
 		TransactionValidationEnabled: true,
 		ChainID:                      flowgo.Emulator,
+		CoverageReportingEnabled:     true,
 	}
```
I think we can get this from command line flag ( default can be true ) I am also totally fine with making it default on without any flag. but if we make config I think making it configurable would be better

m-Peter replied:
Yes, I definitely have to add a command line flag also. But I guess we need this one also, in case the `Blockchain` is utilized programmatically, for example, from another library. In this case the command line flag will not be sufficient I think. Correct me if I'm wrong though.

bluesign replied:
true, you can expose it with a function like `WithCoverageReportingEnabled`  ( example: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/blockchain.go#L368 ) 

then you need to add to `configureBlockchain`: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/server/server.go#L335

then to this `config`: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/server/server.go#L85

and finally: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/cmd/emulator/start/start.go#L42

and 

https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/cmd/emulator/start/start.go#L161

Turtles all the way down :)

m-Peter replied:
Flags, flags everywhere :joy_cat: 
Updated in https://github.com/onflow/flow-emulator/pull/332/commits/8a6d81abc55a31654b1819e8391cec658e2e2af8

bluesign commented:
great work @m-Peter, left few comments

bluesign commented on blockchain.go:
```diff
@@ -445,6 +450,8 @@ func configureFVM(blockchain *Blockchain, conf config, blocks *blocks) (*fvm.Vir
 		fvmOptions...,
 	)
 
+	blockchain.coverageReport = ctx.ReusableCadenceRuntimePool.CoverageReport
+
```
`./blockchain.go:453:61: ctx.ReusableCadenceRuntimePool.CoverageReport undefined (type "github.com/onflow/flow-go/fvm/runtime".ReusableCadenceRuntimePool has no field or method CoverageReport)`

m-Peter replied:
I will have to bump the `flow-go` version, with a release that includes the changes from here: https://github.com/onflow/flow-go/pull/4038

m-Peter commented:
@turbolent I have added a commit (https://github.com/onflow/flow-emulator/pull/332/commits/6a46c114c8aeea46458c580ba7c4082419100917) which attempts to apply the suggestion from https://github.com/onflow/flow-go/pull/4038. It does not work as expected though, I guess I am missing something in the implementation.

![Screenshot from 2023-03-16 13-32-58](https://user-images.githubusercontent.com/1778965/225611708-7a741480-45be-458e-97fc-d2bb19e38fe5.png)

The contract appears on the coverage report, only after the relevant script is executed. And for some reason, the declarations inside the `init()` method, have not been correctly recorded.

![Screenshot from 2023-03-16 13-37-43](https://user-images.githubusercontent.com/1778965/225611892-f61fa10e-0824-4872-8a58-a5f9716625d5.png)

With the previous implementation, just by deploying the contract, it would immediately appear in the coverage report. And the declarations inside the `init()` method, have been correctly recorded.

Do you have any clue as to why this happens? :pray:

turbolent commented on blockchain.go:
```diff
@@ -207,6 +210,7 @@ var defaultConfig = func() config {
 		Logger:                       zerolog.Nop(),
 		TransactionValidationEnabled: true,
 		ChainID:                      flowgo.Emulator,
+		CoverageReportingEnabled:     true,
```
Should this be enabled by default?

m-Peter replied:
Not really, it's up for discussion. On another topic, have you had a look on this https://github.com/onflow/flow-emulator/pull/332#issuecomment-1471840314? I am struggling to find the cause of this, so we can contain the changes in this repository, and discard the `flow-go` companion PR.

m-Peter replied:
Updated default value to `false`

sideninja commented on cmd/emulator/start/start.go:
```diff
@@ -73,6 +73,7 @@ type Config struct {
 	ChainID                   string        `default:"emulator" flag:"chain-id" info:"chain to emulate for address generation. Valid values are: 'emulator', 'testnet', 'mainnet'"`
 	RedisURL                  string        `default:"" flag:"redis-url" info:"redis-server URL for persisting redis storage backend ( redis://[[username:]password@]host[:port][/database] ) "`
 	SqliteURL                 string        `default:"" flag:"sqlite-url" info:"sqlite db URL for persisting sqlite storage backend "`
+	CoverageReportingEnabled  bool          `default:"true" flag:"coverage-reporting" info:"enable Cadence code coverage reporting"`
```
probably better to not be enabled by default?

sideninja replied:
why resolved?

bluesign replied:
sorry I misclicked on this probably

m-Peter replied:
Good point, I will change the default behavior :+1:

m-Peter replied:
Updated the default value to `false`

bluesign commented:
@m-Peter  can be related to reusable runtime fix we talked about.

sideninja commented on coverage_reported_runtime.go:
```diff
@@ -0,0 +1,172 @@
+
+func (crt CoverageReportedRuntime) NewScriptExecutor(
+	script runtime.Script,
+	context runtime.Context,
```
context should be the first argument

sideninja commented:
@m-Peter should we get this merged?

m-Peter commented:
> @m-Peter can be related to reusable runtime fix we talked about.

@bluesign I tried this as well, but the same result, it only collects code coverage reports, upon script/transaction execution. It also misses some parts which are collected upon contract deployment. I am out of ideas on why this is the case, with a custom reusable Cadence runtime pool. I have applied the suggested comments, on the companion PR (https://github.com/onflow/flow-go/pull/4038).

m-Peter commented:
@bluesign Appreciate the help :pray: In this https://github.com/onflow/flow-emulator/pull/332#issuecomment-1471840314 you can see some screenshots with the exact problem.

m-Peter commented:
Yes, as the state of the two PRs, it works as expected. I edited my previous comment, sorry for that :sweat_smile: 

> Note: The problem appears only when we try to contain the changes in this repository, as pattyshack suggested here (https://github.com/onflow/flow-go/pull/4038#discussion_r1137460821). With the companion PR ihttps://github.com/https://github.com/onflow/flow-go/pull/4038, things work as expected.

bluesign commented:
@m-Peter yep it turned out that empty config related. ( the fix we discussed before )

bluesign commented:
```go
config := runtime.Config{
		Debugger:                 blockchain.debugger,
		AccountLinkingEnabled:    true,
		CoverageReportingEnabled: conf.CoverageReportingEnabled,
	}
	coverageReportedRuntime := CoverageReportedRuntime{
		Runtime:        runtime.NewInterpreterRuntime(config),
		CoverageReport: runtime.NewCoverageReport(),
		Environment:    runtime.NewBaseInterpreterEnvironment(config),
	}

	customRuntimePool := reusableRuntime.NewCustomReusableCadenceRuntimePool(
		10,
		func(config runtime.Config) runtime.Runtime {
			return coverageReportedRuntime
		},
	)

	for i := 0; i < 10; i++ {
		customRuntimePool.Return(reusableRuntime.NewReusableCadenceRuntime(coverageReportedRuntime, config))
	}
```

PS: not proud of this code :D

m-Peter commented:
@bluesign Hmm, I was sure that I tried it with the fix from Bastian, since I used the `master` branch. Let me check again, most likely I missed something.

bluesign commented:
@m-Peter seems it is not fixed the custom problem.

m-Peter commented:
My sincere apologies for the numerous `force-push` operations :pray:
After some back and forth with @bluesign, and his excellent debugging skills, I have brought this to what I consider a final version, ready for review. The prerequisite is the PR https://github.com/onflow/flow-go/pull/4126.

sideninja reviewed (changes requested):
https://github.com/onflow/flow-go/pull/4126

bluesign commented:
@m-Peter I am not sure how I feel about [coverage_reported_runtime.go](https://github.com/onflow/flow-emulator/pull/332#diff-a3eb5c3acfb7ec0fb1de7cc95f8750ee021315c8324a875c6a2a6f652b827021) I think @janezpodhostnik and @turbolent knows better but it feels like a bit off to me. 

I feel there should be another way to push coverageReport ? with flow-go fix do we still need this?

m-Peter commented:
@bluesign I personally agree with you, having to create a runtime wrapper, for the sake of coverage report, seems like an extra place where we need to conform with the runtime interface. However, this was suggested from the flow-go team, as an alternative that doesn't need any changes besides this repository. Since you uncovered the bug of config swallow, it turns out that it needs at least that. That's why I have closed the previous companion PR of flow-go (https://github.com/onflow/flow-go/pull/4038). Personally I like the approach that Janez suggested, so that we don't create a runtime wrapper, but we add the method to inject a coverage report in the reusable cadence runtime pool. How do you feel about this alternative?

bluesign commented:
I think it would be best to pass it as runtime.Config:

```go
runtime.Config{
		Debugger:                 blockchain.debugger,
		AccountLinkingEnabled:    true,
		CoverageReportingEnabled: conf.CoverageReportingEnabled,
                CoverageReport: blockchain.coverageReport
}),
```

with config fix, it should be enough without touching the flow-go.  



Btw totally off topic but it would be nice to expose CoverageReport via interface, something like: 

```go
type LineHitReporter interface {
  AddLineHit(location Location, line int) 
}
```

so people can add different coverage modes. But actually what would be better is to put `OnStatement` handler to runtime config. https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#LL147C3-L147C14

This way coverage report doesn't need to use this : https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#L608

can take this function to coverageReport and hook it to OnStatement handler. 
https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#L613-L622

sideninja commented:
@m-Peter what is the status of this PR?

m-Peter commented:
@sideninja The prerequisite PR was merged yesterday (https://github.com/onflow/flow-go/pull/4126), and with it in a released version of `flow-go` we are ready to merge :pray: . Nothing else to add on my side.

m-Peter commented:
I did a rebase on master, to get the changes from https://github.com/onflow/flow-emulator/pull/347. It includes the prerequisite PR.

sideninja commented:
@bluesign can you re-review?

bluesign commented:
@m-Peter I was gonna review but shouldn't we update this to use https://github.com/onflow/flow-emulator/pull/332 ?

m-Peter commented:
@bluesign I think we can first make this available to the developer community, and we can follow up for this change. As it is, it works, and only the internal implementation will change. It will take some time to make all the releases available and update the dependencies.. What do you think?

bluesign commented:
@m-Peter yeah but I think we at least need flow-go PR to merged here ( custom pool fix ) otherwise it breaks the debugger probably.

m-Peter commented:
@bluesign Yes, it was released, and updated on this repository. I have rebased to include it. But if it's not much trouble, you can double check that nothing was broken.

bluesign commented:
> @bluesign Yes, it was released, and updated on this repository. I have rebased to include it. But if it's not much trouble, you can double check that nothing was broken.

sure I will have a look

bluesign commented:
@m-Peter needs rebase to master ( probably on the way you will also need to update the flow-go version, as we don't have the patched one currently it seems )

m-Peter commented:
@bluesign I have rebased after pulling the latest changes from master. It seems to me that we do have the fix though, otherwise, this piece of [code](https://github.com/onflow/flow-emulator/pull/332/files#diff-1fb2891d150a66e1319495e4f10a7ce04855476836a3a6b17f253e312f37d23bR513-R519), would have compilation errors. Any clue?

bluesign commented:
seems all good after rebase, I think we can merge this, then after cadence changes, we can come back to this.
